data preparation
<center><img src=images/image00.png>


# <center> Проект 2: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
import numpy as np

In [ ]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
# параметры подключения удалены

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
# Согласно условий все необходимые таблицы находятся в схеме public 
# Ее указываем перед именем таблицы через точку

query_3_1 = f'''select 
                    count(*) as "Число вакансий"
                from public.vacancies
                '''
answer_3_1 = pd.read_sql_query(query_3_1, connection)
display(answer_3_1.iloc[0, :])


C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\2098150103.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_3_1 = pd.read_sql_query(query_3_1, connection)


Число вакансий    49197
Name: 0, dtype: int64

In [5]:
# ПРОВЕРКА 3_1
vacancies = pd.read_csv('data/vacancies.csv', sep=';')
print('Число вакансий:', vacancies.shape[0])
# display(vacancies)

Число вакансий: 49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''select 
                    count(*) as "Число работодателей"
                from public.employers
                '''
answer_3_2 = pd.read_sql_query(query_3_2, connection)
answer_3_2


C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\3097097887.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_3_2 = pd.read_sql_query(query_3_2, connection)


Число работодателей
0                23501

In [7]:
# ПРОВЕРКА 3_2
employers = pd.read_csv('data/employers.csv', sep=';')
print('Число работодателей:', employers.shape[0])
# display(employers)

Число работодателей: 23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса

query_3_3 = f'''select 
                    count(*) as "Число городов"
                from public.areas
                '''
answer_3_3 = pd.read_sql_query(query_3_3, connection)

display(answer_3_3 .iloc[0, :])

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\604195286.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_3_3 = pd.read_sql_query(query_3_3, connection)


Число городов    1362
Name: 0, dtype: int64

In [9]:
# ПРОВЕРКА 3_3
areas = pd.read_csv('data/areas.csv', sep=';')
print('Число городов:', areas.shape[0])
# display(areas)


Число городов: 1362


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''select 
                    count(*) as "Число сфер деятельности"
                from public.industries
                '''
answer_3_4  = pd.read_sql_query(query_3_4, connection)
display(answer_3_4.iloc[0, :])


C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\3931554579.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_3_4  = pd.read_sql_query(query_3_4, connection)


Число сфер деятельности    294
Name: 0, dtype: int64

In [11]:
# ПРОВЕРКА 3_4
industries = pd.read_csv('data/industries.csv', sep=';')
print('Число сфер деятельности:', industries.shape[0])
#display(industries)

Число сфер деятельности: 294


In [12]:
# Загрузка пятой таблицы (которой нет в вопросах 3.1-3.4)
employers_industries = pd.read_csv('data/employers_industries.csv', sep=';')
#display(employers_industries)

***

## Выводы по предварительному анализу данных
(рассмотрены только данные полученные в вопросах 3.1-3.4 без "влезания внутрь" таблиц)
1. Каждый зарегистрированный работодатель предлагает в среднем 2 вакансии
2. По каждой сфере деятельности вакантно в средней 167 мест
3. В каждом городе в среднем вакантно в среднем 36 мест
4. В каждом городе представлено в среднем 17 работодателей
5. В каждой сфере деятельности имеются предложения в среднем от 80 работодателей


# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [13]:
# текст запроса
# в ответе нужны наименования городов, поэтому присоединяет таблицу areas с городами
query_4_1 = f'''select 
                    a.name,
                    count(v.id) as cnt
                from public.vacancies as v
                join public.areas as a on v.area_id = a.id
                group by a.name 
                order by count(v.id) desc, a.name 
                '''
answer_4_1 = pd.read_sql_query(query_4_1, connection)
display(answer_4_1)


C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\1061644937.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_4_1 = pd.read_sql_query(query_4_1, connection)


name   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764           Шушары     1
765      Электроугли     1
766           Эртиль     1
767             Юрга     1
768               Яя     1

[769 rows x 2 columns]

In [14]:
# ПРОВЕРКА 4_1
answer_4_11 = vacancies.merge(areas,
                            left_on='area_id',
                            right_on='id',
                            sort=True
                            )
answer_4_11.groupby(['name_y']) ['id_x'].agg(['count']).sort_values(by='count', ascending=False)


count
name_y                
Москва            5333
Санкт-Петербург   2851
Минск             2112
Новосибирск       2006
Алматы            1892
...                ...
Надым                1
Навои                1
Навашино             1
Мурмаши              1
Яя                   1

[769 rows x 1 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
# текст запроса
query_4_2 = f'''select 
                    format('Данные о з/п содержатся в %s вакансий', count(id))
                    from public.vacancies
                where (salary_from is not null) OR (salary_to is not null) 
                '''
answer_4_2 = pd.read_sql_query(query_4_2, connection)
display(answer_4_2.iloc[0, 0])


C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\1229318457.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_4_2 = pd.read_sql_query(query_4_2, connection)


'Данные о з/п содержатся в 24073 вакансий'

In [16]:
# ПРОВЕРКА 4_2 с уточнением
# уточение будет заключаться в определении числа анкет совсем без зп, в одним из значений и с вилкой 
# - это для построения диаграммы

# выбираю столбцы с данными о зарплате и проверяю наличие инфы
# Если данных нет, то в этой строке True
null_rows_salary = vacancies[['salary_from', 'salary_to']].isnull()

# Для получения кол-ва анкет без зарплат берем логическое произведение двух логических значений
null_rows_salary['null_salary'] = null_rows_salary['salary_from'] & null_rows_salary['salary_to']
# ячейки True обозначают анкеты, где обе ячейки нули
salary0 = null_rows_salary[null_rows_salary['null_salary'] == True].shape[0]

# Для получения кол-ва анкет с вилкой зарплат берем логическую сумму двух логических значений
null_rows_salary['two_salary'] = (null_rows_salary['salary_from'] | null_rows_salary['salary_to'])
# ячейки False обозначают анкеты, где обе ячейки не нули
salary2 = null_rows_salary[null_rows_salary['two_salary'] == False].shape[0]

# Для получения кол-ва анкет с 1 зарплатой вычитаем из общего кол-ва анкет значения salary0 и salary2
salary1 = answer_3_1.iloc[0,0] - salary0 - salary2

# Для удобства построения создаем DF
answer_4_21 = pd.DataFrame(data=[salary0, salary1, salary2], 
                           columns=['кол-во вакансий'],
                           index=['без зарплат', 'с 1 зарплатой', 'с вилкой'])

print('Данные о кол-ве з/п в вакансиях:')
answer_4_21


Данные о кол-ве з/п в вакансиях:


кол-во вакансий
без зарплат              25124
с 1 зарплатой            13971
с вилкой                 10102

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
# текст запроса
query_4_3 = f'''select 
                round(avg(salary_from), 0) as min_avg_salary,
                round(avg(salary_to), 0) as max_avg_salary
                from public.vacancies
                '''
answer_4_3 = pd.read_sql_query(query_4_3, connection)
display(answer_4_3.iloc[0, :])

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\651432568.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_4_3 = pd.read_sql_query(query_4_3, connection)


min_avg_salary     71065.0
max_avg_salary    110537.0
Name: 0, dtype: float64

In [18]:
# ПРОВЕРКА 4_3
min_avg_salary = round(vacancies['salary_from'].mean(), 0)
max_avg_salary = round(vacancies['salary_to'].mean(), 0)
print('min_avg_salary', min_avg_salary)
print('max_avg_salary', max_avg_salary)

min_avg_salary 71065.0
max_avg_salary 110537.0


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:
# текст запроса
query_4_4 = f'''select 
                schedule,
                employment,
                count(id)
                from public.vacancies
                group by schedule, employment
                order by count(id) desc

                '''
answer_4_4 = pd.read_sql_query(query_4_4, connection)
display(answer_4_4)

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\3213246955.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_4_4 = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

In [20]:
# ПРОВЕРКА 4_4
vacancies.groupby(['schedule', 'employment']) ['id'].agg(['count']).sort_values(by='count', ascending=False)

count
schedule         employment                
Полный день      Полная занятость     35367
Удаленная работа Полная занятость      7802
Гибкий график    Полная занятость      1593
Удаленная работа Частичная занятость   1312
Сменный график   Полная занятость       940
Полный день      Стажировка             569
Вахтовый метод   Полная занятость       367
Полный день      Частичная занятость    347
Гибкий график    Частичная занятость    312
Полный день      Проектная работа       141
Удаленная работа Проектная работа       133
Гибкий график    Стажировка             116
Сменный график   Частичная занятость    101
Удаленная работа Стажировка              64
Гибкий график    Проектная работа        18
Сменный график   Стажировка              12
Вахтовый метод   Проектная работа         2
Сменный график   Проектная работа         1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [21]:
# текст запроса
query_4_5 = f'''select 
                experience,
                count(id)
                from public.vacancies
                group by experience
                order by count(id)

                '''
answer_4_5 = pd.read_sql_query(query_4_5, connection)
display(answer_4_5)

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\2395320505.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_4_5 = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

In [22]:
# ПРОВЕРКА 4_5
vacancies.groupby('experience')['id'].agg(['count']).sort_values(by='count')


count
experience               
Более 6 лет          1337
Нет опыта            7197
От 3 до 6 лет       14511
От 1 года до 3 лет  26152

***

## Выводы по детальному анализу вакансий
1. Всего в базе есть 1362 города (ответ 3_3), а вакансии заявлены из 769 городов (ответ 4_1), т.е. из 57% городов из базы.
2. Число вакансий из Москвы составляет почти 10,5%, а из Санкт-Петербурга почти 6%, в то время как насееление этих городов по официальным даннвым состаялет 9,5 и 4,0 % соответственно (13,3 и 5,6 млн человек из 140 млн). Получается, что в этих городах выакансий относительно больше (т.е. в среднем немного легче найти работу, чем в глубинке). Это вывод при сравнении с официальной статистикой.
Однако этот вывод в реальности неверен - официальные данные не учитывают мигрантов, а по ним ФМС и МВД публикуют весьма разные числа. Значит понимания о кол-ве мигрантов нет вообще.
Возможно, эту составляющую можно было бы отфильтровать, рассматривя вакансии, которые требуют квалификации несколько выше, чем у крьеров или чернорабочих. Однако с каждым годом это будет становится все трудне: постоянно идут новости о том, что миграны становятся врачами, учителями, поступают в ВУЗы и прочее. Т.е. фильтрация по квалификации перестает быть индикатором коренного (учтенного) населения.
3. В 51% объявлений нет сведений о зарплате, хоть какие-то сведения о з/п содержатся в 49% объявлений о вакансиях (ответ 4_2). И лишь в 20,5% есть сведения о вилке зарплат (ответ 4_21).
4. Нижняя граница зарплатной вилки составляет около 71 000 руб., верхняя  - около 110 500 руб. (ответ 4_3).
5. Более половины анкте требуют опыт работы от 1 до 3 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [23]:
# текст запроса
# чтобы получить наименование работодателя присоединем таблицу работодателей employers
query_5_1 = f'''select 
                e.name as name_employer,
                count(v.id)
                from public.vacancies as v
                join public.employers as e ON e.id = v.employer_id
                group by 1
                order by 2 desc, 1 asc
                '''
# Вместо .head(5) в запросе можно было последней строкой написать limit 5
answer_5_1 = pd.read_sql_query(query_5_1, connection)
display(answer_5_1.head(5))
# Смотрим первую и последнюю (пятую) строки из результата

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\1211643596.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_1 = pd.read_sql_query(query_5_1, connection)


name_employer  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

In [24]:
# ПРОВЕРКА 5_1
answer_5_11 = vacancies.merge(employers,
                            left_on='employer_id',
                            right_on='id',
                            sort=True
                            )
answer_5_11 = answer_5_11.groupby(['name_y'])['id_y'].agg(['count']).sort_values(by='count', ascending=False)
display(answer_5_11.head())

count
name_y              
Яндекс          1933
Ростелеком       491
Тинькофф         444
СБЕР             428
Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.

In [25]:
# текст запроса
query_5_2 = f'''select
                        a.name,
                        count(e.id) "работодатели",
                        count(v.id) "вакансии"
                from areas a
                left join vacancies v on a.id = v.area_id
                left join employers e on a.id = e.area
                group by 1
                HAVING count(v.id) = 0 
                order by 2 desc
                limit 1
                '''
answer_5_2 = pd.read_sql_query(query_5_2, connection)
display(answer_5_2)


C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\612937001.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_2 = pd.read_sql_query(query_5_2, connection)


name  работодатели  вакансии
0  Россия           410         0

### <center> Примечание. </center>
Это решение (query_5_2) работет только, если мы ищем строки с нулевыми вакансиями  
count(v.id) = 0   

Однако стоит из запроса убрать строки  
HAVING count(v.id) = 0   
limit 1  

видим, что решение не универсально: для строк, где вакансий не 0, число работодателей и вакансий явно вычислено неверно (bcxbckztncz миллиноны).    
Нашел универсальное решение - см ниже, и в нём, убирая строки having и limit, получим "разумные" значения для любого числа вакансий.  
Идея в том, что один из запросов ("регион - вакансии" или "регион - работодатели") нужно представить "временной" таблицей и приставить ее ко второй "не времненной".  
Для проверки на одном скрине представлены "простые" запросы "регион - вакансии", "регион - работодатели" (я их уже не сохранял) и "полный" запрос (query_5_25) с их результатами (без строк having и limit).
<center><img src=images/image52.png>


In [26]:
# универсальное решение - для любого числа вакансий и работодатедей
query_5_25 = f'''
    with areas_vacancies as
        (select
            a.name,
            a.id,
            count(v.name) as cnt_v
        from public.areas a
        left join public.vacancies v ON v.area_id = a.id
        group by 1, 2
        order by 2 desc
        )

    select
        a_v.name as "регоин",
        a_v.cnt_v as "вакансии",
        count(e.id) as "работодатели"
    from areas_vacancies as a_v
    left join public.employers e ON e.area = a_v.id 
    group by 1, 2
    having a_v.cnt_v = 0
    order by 3 desc, 2 desc, 1
    limit 1
    '''
answer_5_25 = pd.read_sql_query(query_5_25, connection)
display(answer_5_25)


C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\1896215869.py:25: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_25 = pd.read_sql_query(query_5_25, connection)


регоин  вакансии  работодатели
0  Россия         0           410

In [27]:
# ПРОВЕРКА 5_2
# созадем таблицу "вакансий по региону"
areas_vacancies = areas.merge(vacancies,
                            how='left',
                            left_on='id',
                            right_on='area_id'
                            )
areas_vacancies = areas_vacancies.groupby(by='name_x', as_index=False)['employer_id'].agg(['count']).sort_values(by='count', ascending=False)
#display(areas_vacancies)

# созадем таблицу "работодателей по региону"
areas_employers = areas.merge(employers,
                            how='left',
                            left_on='id',
                            right_on='area'
                            )
areas_employers = areas_employers.groupby(by='name_x', as_index=False)['id_y'].agg(['count']).sort_values(by='count', ascending=False)
#display(areas_employers)

# объединяем таблицы
answer_5_21 = areas_vacancies.merge(areas_employers,
                            left_on='name_x',
                            right_on='name_x',
                            )
answer_5_21 = answer_5_21.rename(columns={'name_x':'регион','count_x':'вакансии','count_y':'работодатели'})

# Фильтрация строк c нулевыми вакансиями и сортировка по убыванию
answer_5_21 = answer_5_21[answer_5_21['вакансии'] == 0].sort_values(by='работодатели', ascending=False)
display(answer_5_21.iloc[0, :])


регион          Россия
вакансии             0
работодатели       410
Name: 855, dtype: object

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [28]:
# текст запроса
# Только имена работодателей без их id оставлять нельзя - у некторых
# работодателей наименования полностью совпадают:
# проверил, что чиисло уникальных наименований работодателей 14766, 
# т.е немного меньше, чем число уникальных id работодателей 14906
# А оставлять только id "некрасиво" - человек лучше читает названия

query_5_3 = f'''select 
                v.employer_id,
                e.name,
                count(distinct v.area_id) as "Регионы с объявлениями"
                from public.vacancies as v
                join public.employers as e ON v.employer_id = e.id
                group by 1, 2
                order by 3 desc, 1
                '''
answer_5_3 = pd.read_sql_query(query_5_3, connection)
display(answer_5_3)

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\2874349380.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_3 = pd.read_sql_query(query_5_3, connection)


employer_id                               name  Регионы с объявлениями
0             1740                             Яндекс                     181
1             2748                         Ростелеком                     152
2          5724811                         Спецремонт                     116
3          5130287             Поляков Денис Иванович                      88
4          3682876                          ООО ЕФИН                       71
...            ...                                ...                     ...
14901      8901903                            ЗАРНИЦА                       1
14902      8903436  Kiber One, школа программирования                       1
14903      8903831                        EasyDrive24                       1
14904      8904282   Центр Цифровых Технологий - МФТИ                       1
14905      8906189              Ходыч Роман Сергеевич                       1

[14906 rows x 3 columns]

In [29]:
# ПРОВЕРКА 5_3
answer_5_31 = vacancies.merge(employers,
                            left_on='employer_id',
                            right_on='id',
                            sort=True
                            )
answer_5_31 = answer_5_31.groupby(['employer_id', 'name_y'])['area_id'].agg(['nunique']).sort_values(by=['nunique', 'employer_id'], 
                                                                                             ascending=[False, True])
answer_5_31 = answer_5_31.rename(columns={'nunique': 'Регионы с объявлениями'})
answer_5_31 


Регионы с объявлениями
employer_id name_y                                                   
1740        Яндекс                                                181
2748        Ростелеком                                            152
5724811     Спецремонт                                            116
5130287     Поляков Денис Иванович                                 88
3682876     ООО ЕФИН                                               71
...                                                               ...
8901903     ЗАРНИЦА                                                 1
8903436     Kiber One, школа программирования                       1
8903831     EasyDrive24                                             1
8904282     Центр Цифровых Технологий - МФТИ                        1
8906189     Ходыч Роман Сергеевич                                   1

[14906 rows x 1 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [30]:
# текст запроса
# Чтобы не исчезли строки со значением NULL из "правой" таблицы,
# применяю тип присоединения FULL
query_5_4 = f'''select 
                e.id,
                e.name,                
                count(ei.industry_id) as "Число сфер деятельности"
                from public.employers as e
                full join public.employers_industries as ei ON ei.employer_id = e.id
                group by 1, 2
                having count(ei.industry_id) = 0
                order by 3 desc, 1
                '''
answer_5_4 = pd.read_sql_query(query_5_4, connection)
# display(answer_5_4)
print(answer_5_4.shape[0], 'предприятий не указали сферу деятельности')

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\3331191730.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_4 = pd.read_sql_query(query_5_4, connection)


8419 предприятий не указали сферу деятельности


In [31]:
# ПРОВЕРКА 5_4
answer_5_41 = employers.merge(employers_industries,
                            how='outer',
                            left_on='id',
                            right_on='employer_id',
                            sort=True
                            )

answer_5_41 = answer_5_41.groupby(['id', 'name'])['industry_id'].agg(['nunique']).sort_values(by=['nunique', 'id'], ascending=[False, True])
# Фильтруем строки только с нулевым значением числа сфер деятельности
answer_5_41  = answer_5_41 [answer_5_41['nunique'] == 0]
answer_5_41
# Далее видим размерность таблицы, 
# но можно по желанию впииать строку с shape[0] - как ячейкой выше.

nunique
id      name                                      
960     ОМС                                      0
1304    Luxoft                                   0
1811    Упаковка                                 0
2053    ФПК Гарант-Инвест                        0
3204    ГК ТСС                                   0
...                                            ...
8903436 Kiber One, школа программирования        0
8903831 EasyDrive24                              0
8904282 Центр Цифровых Технологий - МФТИ         0
8905803 Ультра 100                               0
8906189 Ходыч Роман Сергеевич                    0

[8419 rows x 1 columns]

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [32]:
# текст запроса
# Запрос почти аналогичен:
# - исправлено число сфер (having)
# - исправлен столбец сортировки
# - выведена третья строка
query_5_5 = f'''select 
                e.id,
                e.name,                
                count(ei.industry_id) as "Число сфер деятельности"
                from public.employers as e
                full join public.employers_industries as ei ON ei.employer_id = e.id
                group by 1, 2
                having count(ei.industry_id) = 4
                order by 2, 1
                offset 2 
                limit 1
                '''
answer_5_5 = pd.read_sql_query(query_5_5, connection)
display(answer_5_5)


C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\4180993125.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_5 = pd.read_sql_query(query_5_5, connection)


id  name  Число сфер деятельности
0  64174  2ГИС                        4

In [33]:
# ПРОВЕРКА 5_5
answer_5_51 = employers.merge(employers_industries,
                            how='outer',
                            left_on='id',
                            right_on='employer_id',
                            sort=True
                            )

answer_5_51 = answer_5_51.groupby(['id', 'name'])['industry_id'].agg(['nunique']).sort_values(by=['nunique', 'name'], ascending=[False, True])
# Фильтруем строки только с 4 сферами деятельности
answer_5_51 = answer_5_51[answer_5_51['nunique'] == 4]
answer_5_51.iloc[3,:]
# Далее видим размерность таблицы, 
# но можно по желанию впииать строку с shape[0] - как ячейкой выше.

nunique    4
Name: (64174, 2ГИС), dtype: int64

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [34]:
# текст запроса
query_5_6 = f'''select 
                i.name,
                count(ei.employer_id)
                from public.employers_industries as ei
                join public.industries as i ON i.id = ei.industry_id
                group by i.name
                having i.name = 'Разработка программного обеспечения'
                '''
answer_5_6 = pd.read_sql_query(query_5_6, connection)
display(answer_5_6)

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\3552972573.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_6 = pd.read_sql_query(query_5_6, connection)


name  count
0  Разработка программного обеспечения   3553

In [35]:
# ПРОВЕРКА 5_6
answer_5_61 = employers_industries.merge(industries,
                            left_on='industry_id',
                            right_on='id',
                            sort=True
                            )
answer_5_61 = answer_5_61.groupby(['name'], as_index=False)['employer_id'].agg(['count']).sort_values(by=['count'], ascending=False)
answer_5_61 = answer_5_61[answer_5_61['name'] == 'Разработка программного обеспечения']
answer_5_61


name  count
188  Разработка программного обеспечения   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [36]:
# код для получения списка городов-милионников
# 1 - загрузка модулей
import requests
from bs4 import BeautifulSoup
import html.parser

# 2- подключение через мой User-Agent
headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 6.3; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36'}
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
response = requests.get(url, headers=headers)
# Проверяю корректность соединения
response


<Response [200]>

In [37]:
# код для получения списка городов-милионников (проддолжение)
# # 3 - загрузка страницы и таблицы номер 0 (первой)
soup = BeautifulSoup(response.content, "html.parser")
table = soup.find_all('table')[0]
rows = table.find_all('tr')

# 4 - из таблицы забираем только список городов - т.е. столбец №1 (второй)
city_list = []
for tr in rows:
    td = tr.find_all('td')[1:]
    for i in td[0:1]:
        city_list.append(i.text)
#city_list

# 5 - "готовим" список для ссылки на него из запроса SQL:
# а именно составляем из него строку с запятыми в качестве разделителей
city_sql = ", ".join([f"'{i}'" for i in city_list])
city_sql


"'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград'"

In [38]:
# текст запроса
# В этом запросе:
# - заготовлена "временная таблица" employers_vacancies, которая для каждого id региона (все регионы) 
# показывает кол-во вакансий от компании Яндекс тольк во городах из city_sql. 
# - id регионов через таблицу areas заменяем на их названия.
# - заготавливаем строку "Total" для суммарного числа вакансий
# После этого создаем DF со списком вакансий по все регионам

query_5_7 = f'''with employers_vacancies as
                (select 
                    v.area_id as area,
                    e.name,
                    count(v.id) as count
                from public.vacancies as v
                join public.employers as e ON v.employer_id = e.id
                group by 1, 2
                having e.name = 'Яндекс' 
                )
            
                select 
                a.name as "регион",
                e_v.count as "число вакансий"
                from employers_vacancies e_v
                join public.areas as a ON a.id = e_v.area
                where a.name in ({city_sql})       

                union
                select
                'Total',
                sum(e_v.count)
                from employers_vacancies e_v
                join public.areas as a ON a.id = e_v.area
                where a.name in ({city_sql})     

                order by 2 desc
                 '''
answer_5_7 = pd.read_sql_query(query_5_7, connection)
display(answer_5_7)

# Получили 181 строку для Яндекса, а в вопросе 5.3 как раз нашли, что Янлекс имеет вакансии по 181 региону.

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\2520421988.py:37: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_7 = pd.read_sql_query(query_5_7, connection)


регион  число вакансий
0             Total           485.0
1            Москва            54.0
2   Санкт-Петербург            42.0
3      Екатеринбург            39.0
4   Нижний Новгород            36.0
5       Новосибирск            35.0
6           Воронеж            32.0
7         Краснодар            30.0
8            Самара            26.0
9               Уфа            26.0
10            Пермь            25.0
11           Казань            25.0
12   Ростов-на-Дону            25.0
13        Волгоград            24.0
14       Красноярск            23.0
15        Челябинск            22.0
16             Омск            21.0

In [39]:
# ПРОВЕРКА_5_7
# Собираем регионы, где есть ваакансии Яндекса
employers_vacancies = vacancies.merge(employers,
                            left_on='employer_id',
                            right_on='id'
                            )
employers_vacancies = employers_vacancies.groupby(['area_id','name_y'], as_index=False)['area_id'].agg(['count']).sort_values(by=['count'], ascending=False)
employers_vacancies = employers_vacancies[employers_vacancies['name_y']=='Яндекс']

# Вместо id регионов записываем их наименования
yandex_vacancies = employers_vacancies.merge(areas,
                            left_on='area_id',
                            right_on='id'
                            )
yandex_vacancies = yandex_vacancies.loc[:, ['name','count']]
#display(yandex_vacancies)

# Для отбора регионов готовим DF регионов 
city_df = pd.DataFrame(city_list,
                       columns=['регион'])
#print(city_df)

# склеивам таблицы
answer_5_71 = city_df.merge(yandex_vacancies,
                            left_on='регион',
                            right_on='name'
                            )
answer_5_71 = answer_5_71.loc[:, ['регион','count']]
answer_5_71= answer_5_71.rename(columns={'count': 'число вакансий'})
x = answer_5_71['число вакансий'].sum()

answer_5_71.loc[16,:] = ['Total', x]

display(answer_5_71)



регион  число вакансий
0            Москва            54.0
1   Санкт-Петербург            42.0
2       Новосибирск            35.0
3      Екатеринбург            39.0
4            Казань            25.0
5        Красноярск            23.0
6   Нижний Новгород            36.0
7         Челябинск            22.0
8               Уфа            26.0
9         Краснодар            30.0
10           Самара            26.0
11   Ростов-на-Дону            25.0
12             Омск            21.0
13          Воронеж            32.0
14            Пермь            25.0
15        Волгоград            24.0
16            Total           485.0

In [40]:
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.graph_objs as go

In [41]:
# Круговая диаграмма
fig = px.pie(answer_5_71.iloc[:15],
             names='регион',
             values='число вакансий',
             title="Вакансии компании Яндекс по регионам"
             )

fig.update_layout(title_x=0.4, 
                  showlegend=True,
                  autosize=False, 
                  width=800, 
                  height=600,

                  margin=dict(
                        l=50,    # left margin
                        r=50,    # right margin
                        b=50,    # bottom margin
                        t=50,    # top margin
                       #pad=1    # заполнение между графиком и полем
                        )
                )
#fig.write_html('images/image71.html')
fig.show()

***

## Выводы по анализу работодателей.
1. У компании Яндекс в 5 раз больше вакансий, чем у других компаний с наибоьшим числом вакансий. Яндекс расширяется ? Или большая текучка?  
2. Работодатели, указавшие регион "Россия" не предоставляют вакансий.   
3. Лидерами по числу объявлений в разных регионах являюся Яндекс и Ростелеком. У них наибольшее кол-во возникакющих региональных отделений? Или наибольшая текучка?.  
4. Почти 36% (более трети) работодателей (8419 из 23501) не указали ни одной сферы деятельности.  
5. Около 5% работодателей (1136) указали 4 сферы деятельности.  
6. Разработка ПО указана сферой деятельности у 15% работодателей (3553).  Очень большая доля!!! Бурное развитие отрасьи....
7. 


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [42]:
# текст запроса
query_6_1 = f'''
    select
        count(v.id) as "Кол-во вакансий, связанных с даннымы"
    from public.vacancies v
    where lower(v.name) LIKE '%данн%' OR lower(v.name) LIKE '%data%'
    '''

answer_6_1 = pd.read_sql_query(query_6_1, connection)
answer_6_1

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\129058213.py:9: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Кол-во вакансий, связанных с данны
0                                1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [43]:
# текст запроса
# в столбце названий вакансий проверяем наличие заданных сочетаний символов
# и одновремненно фильтруем столбец опыт (experience) или наименование 
query_6_2 = f'''
    select
        count(v.id) as "Кол-во вакансий для junior DS"
    from public.vacancies as v
    where (
            lower(v.name) like '%data scientist%'
            OR lower(v.name) like '%data science%'
            OR lower(v.name) like '%исследователь данных%'
            OR (lower(v.name) like '%ml%' AND lower(v.name) NOT like '%html%')
            OR lower(v.name) like '%machine learning%'
            OR lower(v.name) like '%машинн%обучен%'
            )
        AND 
            (
            lower(v.name) like '%junior%'
            OR lower(v.experience) like '%нет опыта%'
            OR lower(v.experience) like '%стажировка%'
            )
    '''

answer_6_2 = pd.read_sql_query(query_6_2, connection)
answer_6_2

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\2713053740.py:24: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Кол-во вакансий для junior DS
0                             51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [44]:
# текст запроса
# берем первую группу условий (до слова AND) из предыдущего ответа
# и дописываем условие по навыкам (столбец key_skills)
query_6_3 = f'''
    select
        count(v.id) as "Кол-во вакансий для DS с навыками"
    from public.vacancies as v
    where (
            lower(v.name) like '%data scientist%'
            OR lower(v.name) like '%data science%'
            OR lower(v.name) like '%исследователь данных%'
            OR lower(v.name) like '%machine learning%'
            OR (lower(v.name) like '%ml%' AND lower(v.name) NOT like '%html%')
            OR lower(v.name) like '%машинн%обучен%'
            )
        AND
            (
            lower(v.key_skills) LIKE '%sql%'
            OR lower(v.key_skills) LIKE '%postgres%'
            )
    '''
answer_6_3 = pd.read_sql_query(query_6_3, connection)
answer_6_3 

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\240130488.py:22: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Кол-во вакансий для DS с навыками
0                                229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [45]:
# текст запроса
# изменения только после слова AND
query_6_4 = f'''
    select
        count(v.id) as "Кол-во вакансий для DS с Python"
    from public.vacancies as v
    where (
            lower(v.name) like '%data scientist%'
            OR lower(v.name) like '%data science%'
            OR lower(v.name) like '%исследователь данных%'
            OR lower(v.name) like '%machine learning%'
            OR (lower(v.name) like '%ml%' AND lower(v.name) NOT like '%html%')
            OR lower(v.name) like '%машинн%обучен%'
            )
        AND
           lower(v.key_skills) LIKE '%python%'
         
    '''
answer_6_4 = pd.read_sql_query(query_6_4, connection)
answer_6_4 

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\179985446.py:19: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Кол-во вакансий для DS с Python
0                              357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [46]:
# текст запроса
# В столбце key_skills отдельные навыки разделены символом "\t" - и интерпретируется как один символ "табуляции", 
# поэтоту число навыков считаем так: 
# исключаем все такие разделители, строка становится короче на число таких разделителей, 
# а число "слов" больше числа разделителей на 1.
# При этом для avg автоматически null учитывается, как 0

# При фильтрации используем первую часть (до слова AND) фильтрации - определение всех вакансий DS

query_6_5 = f'''
    select
            round(avg(length(key_skills) - length(REPLACE(key_skills, '\t', '')) +1 ), 2) as "Среднее число навыков"
    from public.vacancies v
    where   (
            lower(v.name) like '%data scientist%'
            OR lower(v.name) like '%data science%'
            OR lower(v.name) like '%исследователь данных%'
            OR lower(v.name) like '%machine learning%'
            OR (v.name like '%ML%' AND lower(v.name) NOT like '%html%')
            OR lower(v.name) like '%машинн%обучен%'
            )
        order by 1
    '''

answer_6_5 = pd.read_sql_query(query_6_5, connection)
answer_6_5

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\281081703.py:25: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Среднее число навыков
0                   6.41

In [47]:
# результат запроса

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [48]:
# текст запроса
query_6_6 = f'''
    select
        'От 3 до 6 лет' as "Опыт работы",
        round(AVG(COALESCE((v.salary_to + v.salary_from)/2, v.salary_to, v.salary_from))) as "Средняя зарплата для DS"
    from public.vacancies as v
    where (
            lower(v.name) like '%data scientist%'
            OR lower(v.name) like '%data science%'
            OR lower(v.name) like '%исследователь данных%'
            OR lower(v.name) like '%machine learning%'
            OR (v.name like '%ML%' AND lower(v.name) NOT like '%html%')
            OR lower(v.name) like '%машинн%обучен%'
        )
            AND (v.salary_to IS NOT NULL OR v.salary_from IS NOT NULL)
            AND v.experience = 'От 3 до 6 лет'
    '''

answer_6_5 = pd.read_sql_query(query_6_6, connection)
answer_6_5

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\2011295336.py:19: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Опыт работы  Средняя зарплата для DS
0  От 3 до 6 лет                 243115.0

***

## Выводы по предметному анализу.
1. Количество вакансий, связанных с работой с данными, составляет 1771, что составляет более 3,5% от всех вакансий. Имеет место высокий спрос на специалистов в этой области.
2. Существует 51 вакансия для начинающего дата-сайентиста – это 3% от вакансий, связанных с работой с данными. Это означает, что существуют возможности для новичков в этой области начать свою карьеру.
3. Количество вакансий, связанных с работой с данными, где SQL или postgre указаны в качестве ключевых навыков, составляет 229 – 13% от общего числа. Работодатель достаточно часто запрашивает эти навыки.
4. Количество вакансий, где Python указывается в качестве ключевого навыка, составляет 357 – около 20%. Т.е. знание Python востребовано в работе, и видимо считается одним из основных инструментов.
5. Среднее количество ключевых навыков составляет более 6. Значит, работодатель в этой области требует специалистов с широким спектром навыков и компетенций. 
6. Средняя зарплата для специалиста с опытом работы от 3 до 6 лет составляет 243115 руб. Весьма высокая, особенно для регионов - даже для 2025 г.!!!


# Общий вывод по проекту

## Общие выводы.
1. Предоставленная выборка содержит:
- 49197 вакансий, 
- данные о 23501 работодателе, из которых 14906 предлагают вакансии,
- информацию о 1362 регионах по 294 сферам деятельности.
2. Лидерами по количеству вакансий являются мегаполисы.
3. Среднее значение для зарплатной вилки составляет от 71  до 110,5 тыс. руб.
4. Опыт работы от 1 года до 3 лет является наиболее востребованным.
5. Полная занятость является наиболее популярным типом занятости.
6. Компания "Яндекс" занимает первое место по количеству вакансий - 1933. Кроме того она значительно лидирует по числу охваченных вакансиями регионов.
7. Количество вакансий, связанных с работой с данными, составляет 1771. Одним из самых востребованных навыков является владение Python (20%). Среднее количество ключевых навыков для специалистов по работе с данными составляет более 6.
8. Соискатель со средним опытом может рассчитывать в среднем на зарплату в 243115 руб. – весьма высокую даже в 2025 г.

## Возможные направления исследований.
1. Необходимо провести работы по унификации данных. Встречаются вакансии имеющие различное написание (состав символов), но фактически одинаковое значение.
2. По жизненному опыту поисков работы часто работодатель завышает требования – тут требуется доп исследование: опрос получивших работу – а все ли навыки, отраженные в вакансии, оказались востребованы.
3. Желательно провести анализ трендовых вакансий: в целом и в зависимости от типа занятости
4. Рассмотреть распределение зарплат по профессиям
5. Можно попробовать рассмотреть распределение зарплат по регионам России. Но для этого нужно пересмотреть таблицу areas:
 - спарсить список стран и удалить их из таблицы вакансий
 - выделить отдельно Москву, Санкт-Петегрбург, города-миллионники, прочие. Можно еще отдельно попробовать разбить "прочие"+"города-миллионники" на федеральные округа или федеральные регионы, но для этого нужно спарсить (и скорее всего дополнить) списки городов по этим округам (регионам). Тогда получим примерно то, что выдает Росстат.

# Юнит 7. Дополнительные исследования

1. К юниту 3. 
Сколько работодателей из зарегистрированных предлагают вакансии?  
Сравнить с числом работодателей без вакансий.

In [49]:
query_7_1 = f'''select 
                    'с вакансиями' as "Наличие вакансий",
                    count(distinct v.employer_id) as "Число работодателей"                
                from public.vacancies as v
                '''
answer_7_1 = pd.read_sql_query(query_7_1, connection)                
answer_7_1.loc[1] = ['без вакансий', answer_3_2.iloc[0, 0] - answer_7_1.iloc[0,1]]
answer_7_1

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\2992484199.py:6: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Наличие вакансий  Число работодателей
0     с вакансиями                14906
1     без вакансий                 8595

In [50]:
# Круговая диаграмма
fig = px.pie(answer_7_1,
             names='Наличие вакансий',
             values='Число работодателей',
             title="Работодатели")

fig.update_layout(title_x=0.4, 
                  showlegend=True,
                  autosize=False, 
                  width=500, 
                  height=400,
                  barmode='group',
                  margin=dict(
                        l=50,    # left margin
                        r=50,    # right margin
                        b=50,    # bottom margin
                        t=50,    # top margin
                       #pad=1    # заполнение между графиком и полем
                        )
                )
#fig.write_html('images/image71.html')
fig.show()

2. К юниту 4.  
Как зависит зарплата от опыта работы?  
(Использую наработку из задания 6.6)

In [51]:
# текст запроса
query_7_2 = f'''
    select
        v.experience as "Опыт работы",
        round(AVG((COALESCE((v.salary_to + v.salary_from)/2, v.salary_to, v.salary_from))/1000), 1) as "Средняя зарплата, тыс. руб."
    from public.vacancies as v
    group by 1
    order by 2 desc
    '''
answer_7_2  = pd.read_sql_query(query_7_2, connection)
answer_7_2 

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\3278332518.py:10: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Опыт работы  Средняя зарплата, тыс. руб.
0         Более 6 лет                        166.1
1       От 3 до 6 лет                        133.1
2  От 1 года до 3 лет                         76.4
3           Нет опыта                         40.4

In [52]:
# Гистограмма
fig = px.bar(
    answer_7_2,
    x='Опыт работы',
    y='Средняя зарплата, тыс. руб.', 
    title='Зависимость заплаты от опыта',
    text='Средняя зарплата, тыс. руб.'
    )

fig.update_layout(title_x=0.5, 
                  showlegend=True,
                  autosize=False, 
                  width=500, 
                  height=400,
                  
                  margin=dict(
                        l=50,    # left margin
                        r=50,    # right margin
                        b=50,    # bottom margin
                        t=50,    # top margin
                      # pad=3    # заполнение между графиком и полем
                        )
                )
#fig.write_html('images/image72.html')
fig.show()

3. К юниту 5.    
Как распределены зарегистрированные работодатели по кол-ву сфер деятельности?

In [53]:
# текст запроса
# Чтобы не исчезли строки со значением NULL из "правой" таблицы,
# применяю тип присоединения FULL
query_7_3 = f'''with industries_count as
                (select 
                e.id,
                count(ei.industry_id) as count_i
                from public.employers as e
                full join public.employers_industries as ei ON ei.employer_id = e.id
                group by 1
                order by 2 desc, 1)
                
                select
                count_i as "Число указанных сфер деятельности",
                count(ic.id) as "Кол-во работодателей"
                from industries_count as ic
                group by 1
                order by 2 desc
                
                
                '''
answer_7_3 = pd.read_sql_query(query_7_3, connection)
display(answer_7_3)


C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\4151229738.py:22: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Число указанных сфер деятельности  Кол-во работодателей
0                                  0                  8419
1                                  1                  6452
2                                  2                  3639
3                                  3                  2607
4                                  5                  1246
5                                  4                  1138

In [54]:
# Гистограмма
fig = px.bar(
    answer_7_3,
    x='Число указанных сфер деятельности',
    y='Кол-во работодателей', 
    text='Кол-во работодателей',
    title='Распределение работодателей<br>по числу сфер деятельности'
    )

fig.update_layout(title_x=0.5, 
                  showlegend=False,
                  autosize=True, 
                  width=600, 
                  height=400,
                  margin=dict(
                        l=50,    # left margin
                        r=50,    # right margin
                        b=50,    # bottom margin
                        t=70,    # top margin
                      # pad=3    # заполнение между графиком и полем
                        )
                )
#fig.write_html('images/image73.html')
fig.show()

4. К юниту 6.    
Найти топ-10 самых востребованных вакансий в целом по всем вакансиям
Каковы их средние зарплаты?


In [55]:
query_7_4 = f'''with vacancies_count as
                (select
                     name as names, 
                     count(*)
                from public.vacancies
                group by 1
                order by 2 DESC
                limit 10)
                
                select
                names as "Вакансия",
                count as "Число вакансий",
                round(AVG((COALESCE((v.salary_to + v.salary_from)/2, v.salary_to, v.salary_from))/1000), 1) as "Средняя зарплата, тыс. руб."
                from vacancies_count vc
                join public.vacancies as v ON vc.names = v.name
                group by 1, 2
                order by 2 desc
                '''
answer_7_4 = pd.read_sql_query(query_7_4, connection)
display(answer_7_4)

C:\Users\РАФ\AppData\Local\Temp\ipykernel_1700\4184968700.py:19: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Вакансия  Число вакансий  \
0           Системный администратор            1761   
1                    Программист 1С            1153   
2                   Бизнес-аналитик             488   
3  Специалист технической поддержки             457   
4                Системный аналитик             432   
5                          Аналитик             404   
6               Инженер-программист             389   
7                    Программист 1C             360   
8              Frontend-разработчик             242   
9     Инженер технической поддержки             230   

   Средняя зарплата, тыс. руб.  
0                         50.0  
1                        110.2  
2                         92.1  
3                         43.1  
4                        127.3  
5                         61.1  
6                         50.3  
7                         96.9  
8                        115.3  
9                         47.9

In [56]:
# Две гистограммы
fig = px.bar(answer_7_4,
                    x=['Средняя зарплата, тыс. руб.', 'Число вакансий'],
                    y='Вакансия',
                    text='Средняя зарплата, тыс. руб.',
                    title='Число вакансий и средння зарплата для ТОП-10 специальностей<br>(зарплата указана внутри соответствуюющего бара)',
                    orientation='h',
                    labels = {'value': '',
                              'variable':''})

fig.update_layout(title_x=0.5, 
                  showlegend=True,
                  autosize=False, 
                  width=1000, 
                  height=500,
                  margin=dict(
                      l=50,    # left margin
                      r=50,    # right margin
                      b=70,    # bottom margin
                      t=100,    # top margin
                    # pad=3    # заполнение между графиком и полем
                        ),
                  legend=dict(
                      y=0.9,
                      x=0.7
                  )
                )

fig.show()
#fig.write_html('images/image74.html')

## Выводы по доп. исследованиям.
1. Вакансии предоставлены 63,4% работодателей от общего числа зарегистрированных. Хорошо бы исследовать непредоставивших: это новые, давно не предоставляли или недавно предоставляли.
2. Разбег средних зарплат с опытом более 6 лет и без опыта составляет в среднем в 4 раза. Между 1 и 2, 2 и 3 категориями "опытности" зарплата растет почти вдвое, однако с дальнейшим увеличением опыта ее рост резко замедляется. Рекомендация соискателям: не надо долго засижываться на одной специализации.
3. Зарегистрированные работодатели указали не более 5 сфер деятельности. Наименьшее число работодателей (примерно по 5%) указало 4 или 5 сферы деятельности, а порядка 36% не указали ни одной. Возможно, не указавшие  - это предприятия, сфера деятельности которых понятна из названия, либо расположенные в небольших населенных пунктах, где об их сфере деятельности знают многие жители. Этот вопрос можно исследовать.
4. Востребованность не коррелирует с зарплатой. Самый востребованный специалсит получает почти в 2,5 раза меньше, чем самый высокооплачиваемый из топ-10 востребованных. 

In [57]:
connection.close()